### PR TITLE
FIX missing entity filter and better sql request

### PR DESCRIPTION
**Pull request by hregis**
@eldy 
the loop "while" was too long if many bills

### CODE REVIEW

File: htdocs/compta/facture/card.php [hunk 1]
```diff
@@ -2948,15 +2948,12 @@
...
 			print '<div class="tagtr listofinvoicetype"><div class="tagtd listofinvoicetype">';
 			$tmp='<input type="radio" name="type" value="5"' . (GETPOST('type') == 5 && GETPOST('originid') ? ' checked' : '');
-			if ($opt == ('<option value ="0" selected>' . $langs->trans('NoSituations') . '</option>') || (GETPOST('origin') && GETPOST('origin') != 'facture' && GETPOST('origin') != 'commande')) $tmp.=' disabled';
+			if (GETPOST('origin') != 'facture' && GETPOST('origin') != 'commande') $tmp.=' disabled';
 			$tmp.= '> ';
 			$text = $tmp.$langs->trans("InvoiceSituationAsk") . ' ';
-			$text .= '<select class="flat" id="situations" name="situations">';
-			$text .= $opt;
-			$text .= '</select>';
+			$text .= $form->selectSituationInvoices(GETPOST('originid', 'int'), $socid);
```
Comment (eldy) [inline htdocs/compta/facture/card.php:2956]: Where are the < select > around the < option > now ?

Comment (hregis) [reply htdocs/compta/facture/card.php:2956]: @eldy 
if you speak of the "select" tags they have been moved in the `selectSituationInvoices()` function

File: htdocs/core/class/html.form.class.php [hunk 2]
```diff
@@ -3312,55 +3312,70 @@ function formSelectShippingMethod($page, $selected='', $htmlname='shipping_metho
...
 					//Same cycle ?
-			        if ($obj->situation_cycle_ref != $ref) {
+					if ($obj->situation_cycle_ref != $ref)
+					{
 						// Just seen this cycle
-			            $ref = $obj->situation_cycle_ref;
-						//not final ?
-			            if ($obj->situation_final != 1) {
-							//Not prov?
-			                if (substr($obj->facnumber, 1, 4) != 'PROV') {
```
Comment (eldy) [inline htdocs/core/class/html.form.class.php:3343]: Is removing this test on ref of invoice = '(PROV...' done on purpose ?

Comment (hregis) [reply htdocs/core/class/html.form.class.php:3343]: @eldy 
I moved some tests of the "while" loop in the sql query for better performance !

```
$sql = 'SELECT rowid, facnumber, situation_cycle_ref, situation_counter';
$sql.= ' FROM ' . MAIN_DB_PREFIX . 'facture';
$sql.= ' WHERE entity IN ('.getEntity('invoice').')';
$sql.= ' AND situation_counter >= 1';
$sql.= " AND facnumber NOT LIKE '%PROV%'";
$sql.= " AND situation_final != 1";
if (! empty($socid)) {
	$sql.= ' AND fk_soc = ' . $socid;
}
$sql.= ' ORDER BY situation_cycle_ref, situation_counter DESC';
```

**Comment by hregis** [comment]
@eldy ping

File: htdocs/core/class/html.form.class.php (hunk 2 above)
Comment (eldy) [reply htdocs/core/class/html.form.class.php:3343]: It is not clear for me what is fixing the bug or not. Also when commiting into already released branch, we must reduce the PR to the lowest change as possible, even if code can be better or optimize, we must try to change the minimal code the fix the bug and only fix the bug to reduce side effect.
Can you make a PR that contains only the bug fix of missing filter on entity ?

Comment (hregis) [reply htdocs/core/class/html.form.class.php:3343]: @eldy 
ok, look this https://github.com/Dolibarr/dolibarr/pull/10870

**Pull request closed without merging**